jupyter/d2l-java/chapter_linear-networks/linear-regression-scratch.ipynb

```diff
@@ -0,0 +1,698 @@
+    ":numref:`chap_optimization`.\n",
+    "\n",
+    "Note: We can replace `linreg` and `squaredLoss` with any net or loss function respectively\n",
+    "and still keep the same training structure shown here."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "12"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import ai.djl.training.GradientCollector;\n",
+    "import ai.djl.engine.Engine;\n",
+    "\n",
+    "float lr = 0.03f;  // Learning rate\n",
+    "int numEpochs = 3;  // Number of iterations\n",
+    "\n",
+    "for (int epoch = 0; epoch < numEpochs; epoch++) {\n",
+    "    // Assuming the number of examples can be divided by the batch size, all\n",
+    "    // the examples in the training dataset are used once in one epoch\n",
+    "    // iteration. The features and tags of minibatch examples are given by X\n",
+    "    // and y respectively\n",
+    "    for (Batch batch : dataset.getData(manager)) {\n",
+    "        NDArray X = batch.getData().head();\n",
+    "        NDArray y = batch.getLabels().head();\n",
+    "        \n",
+    "        // Attach Gradients\n",
+    "        for (NDArray param : params) {\n",
+    "            param.attachGradient();\n",
+    "        }\n",
+    "        \n",
+    "        try (GradientCollector gc = Engine.getInstance().newGradientCollector()) {\n",
+    "            // Minibatch loss in X and y\n",
+    "            NDArray l = squaredLoss(linreg(X, params.get(0), params.get(1)), y);\n",
+    "            gc.backward(l);  // Compute gradient on l with respect to w and b\n",
+    "            sgd(params, lr, batchSize);  // Update parameters using their gradient\n",
+    "        }\n",
+    "        \n",
+    "        batch.close();\n",
+    "    }\n",
+    "    NDArray trainL = squaredLoss(linreg(features, params.get(0), params.get(1)), labels);\n",
+    "    System.out.printf(\"epoch %d, loss %f\\n\", epoch + 1, trainL.mean().getFloat());\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this case, because we synthesized the data ourselves,\n",
+    "we know precisely what the true parameters are. \n",
+    "Thus, we can evaluate our success in training \n",
+    "by comparing the true parameters\n",
+    "with those that we learned through our training loop. \n",
+    "Indeed they turn out to be very close to each other."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "13"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "float[] w = trueW.sub(params.get(0).reshape(trueW.getShape())).toFloatArray();\n",
+    "System.out.printf(\"Error in estimating w: [%f %f]\\n\", w[0], w[1]);\n",
+    "System.out.print(\"Error in estimating b: \");\n",
+    "System.out.println(trueB - params.get(1).getFloat());"
```

Comment:
I wrote this slightly weirdly since if I use printf as my last statement, then a `java.io.PrintStream` shows up in the output in addition to the printed statement whereas if I use println, nothing else shows up.

This just makes the output cleaner.